proposal/prefixed-parent-selector.md

```diff
@@ -128,20 +136,22 @@ To execute a style rule `rule`:
 
 * If there is a current style rule:
 
-  * If `selector` contains one or more `ParentSelector`:
+  * If `selector` one or more parent selectors, replace each parent selector with
+    the current style rule's selector:
+
+    * If the result of the replacement(s) is not a syntactically valid
+      [complex selector][], throw an error.
```

Comment:
This effectively requires re-parsing the entire selector, which is relatively expensive. Writing it this way also means that a reader has to manually think through all the cases that might be invalid, making it easier to misinterpret the spec.

I think we can much more explicitly and efficiently detect whether the parent is a valid replacement by just enumerating the types of selectors that are not valid when used with prefixes and/or suffixes.

Reply:
If it helps I think the logic is something like:

For each Parent Selector Complex selector, a parent prefix must be compatible with the first CompoundSelector's first component and the suffix must be compatible with the last Compound selector's last component.

- A suffix is compatible with ID, Type, Class, or Pseudo without arguments 
- prefix is compatible with  type selectors or as a type selector itself (I think).

Sorry for jumping in here unannounced, i've been writing a little Sass-Like preprocessor in postcss and PEG.js so have been thinking through the grammar of Sass more than one might normally, and thought it may be helpful to share!

Reply:
Contributions are totally welcome!

I think a prefixed parent selector should be compatible with both type selectors (as long as they aren't universal) *and* what the spec refers to as "subclass selcetors". `.foo { a& { ... } }` and `foo { bar-& { ... } }` use a prefixed parent selector for different purposes, but I think both purposes are valid.

Reply:
Sounds good.  

I think we'll wait for your input on the `[ <pseudo-element-selector> <pseudo-class-selector>* ]*` part of the compound selector specification before starting this though. The number of allowed combinations needing to be listed out depends on whether or not we support things like `::before&:hover`.

Reply:
Also from the conversation it sounds like we want to support things like this:
```
.foo {
    &-bar {
        ...
    }
}
```
to output
```
.foo-bar { ... }
```
(Note that this appears to already work at https://www.sassmeister.com/)

We could also support this:
```
foo {
    .&-bar {
        ...
    }
}
```
also to output
```
.foo-bar { ... }
```

And this:
```
.foo {
    div&-bar {
        ...
    }
}
```
could turn into this:
```
div.foo-bar { ... }
```

I can see these being useful for certain CSS methodologies, but they will add some complexity to the semantics for determining if the current style rule is compatible with the parent selector's location.  For example in the first case the string with the parent selector in it changes from what appears to be a `<ident-token>` (`-bar`) to a `<class-selector>` (`.foo-bar`) after the substitution.

Reply:
I think we should not support `.&-bar` but we should support `div&-bar`. The former, as you say, adds a tremendous amount of syntactic complexity for very little gain. The latter is a natural consequence of allowing both prefixes and suffixes on parent selectors (although I would personally consider it questionable style).

Reply:
`.&-bar` should be disallowed by the syntax I proposed above because the parent selector is in the middle of a class selector.

If we do allow things like `div&-bar` the semantics get a lot simpler because they most just have to check for things like making sure a type selector doesn't end up with two namespaces.